<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><style>body {
  max-width: 980px;
  margin: 16px auto;
}

body .markdown-body
{
  padding: 45px;
}

@font-face {
  font-family: fontawesome-mini;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAABE0AA8AAAAAHWwAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABHU1VCAAABWAAAADsAAABUIIslek9TLzIAAAGUAAAAQwAAAFY3d1HZY21hcAAAAdgAAACqAAACOvWLi0FjdnQgAAAChAAAABMAAAAgBtX/BGZwZ20AAAKYAAAFkAAAC3CKkZBZZ2FzcAAACCgAAAAIAAAACAAAABBnbHlmAAAIMAAABdQAAAjkYT9TNWhlYWQAAA4EAAAAMwAAADYQ6WvNaGhlYQAADjgAAAAfAAAAJAc6A1pobXR4AAAOWAAAACAAAAA0Kmz/7mxvY2EAAA54AAAAHAAAABwQPBJubWF4cAAADpQAAAAgAAAAIAEHC/NuYW1lAAAOtAAAAYQAAALxhQT4h3Bvc3QAABA4AAAAfgAAAMS3SYh9cHJlcAAAELgAAAB6AAAAhuVBK7x4nGNgZGBg4GIwYLBjYHJx8wlh4MtJLMljkGJgYYAAkDwymzEnMz2RgQPGA8qxgGkOIGaDiAIAJjsFSAB4nGNgZHZmnMDAysDAVMW0h4GBoQdCMz5gMGRkAooysDIzYAUBaa4pDA4Pwz+yMwf9z2KIYg5imAYUZgTJAQDcoQvQAHic7ZHNDYJAFIRnBXf94cDRIiyCKkCpwFCPJ092RcKNDoYKcN4+EmMPvpdvk539zQyAPYBCXEUJhBcCrJ5SQ9YLnLJe4qF5rdb+uWPDngNHTkta101pNyWa8lMhn6xx2dqUnW4q9YOIhAOOeueMSgsR/6ry+P7O5s6xVNg4chBsHUuFnWNJ8uZYwrw7chrsHXkODo7cB0dHOYCTY8kv0VE2WJKD6gOlWjsxAAB4nGNgQAMSEMgc9D8LhAESbAPdAHicrVZpd9NGFB15SZyELCULLWphxMRpsEYmbMGACUGyYyBdnK2VoIsUO+m+8Ynf4F/zZNpz6Dd+Wu8bLySQtOdwmpOjd+fN1czbZRJaktgL65GUmy/F1NYmjew8CemGTctRfCg7eyFlisnfBVEQrZbatx2HREQiULWusEQQ+x5ZmmR86FFGy7akV03KLT3pLlvjQb1V334aOsqxO6GkZjN0aD2yJVUYVaJIpj1S0qZlqPorSSu8v8LMV81QwohOImm8GcbQSN4bZ7TKaDW24yiKbLLcKFIkmuFBFHmU1RLn5IoJDMoHzZDyyqcR5cP8iKzYo5xWsEu20/y+L3mndzk/sV9vUbbkQB/Ijuzg7HQlX4RbW2HctJPtKFQRdtd3QmzZ7FT/Zo/ymkYDtysyvdCMYKl8hRArP6HM/iFZLZxP+ZJHo1qykRNB62VO7Es+gdbjiClxzRhZ0N3RCRHU/ZIzDPaYPh788d4plgsTAngcy3pHJZwIEylhczRJ2jByYCVliyqp9a6YOOV1WsRbwn7t2tGXzmjjUHdiPFsPHVs5UcnxaFKnmUyd2knNoykNopR0JnjMrwMoP6JJXm1jNYmVR9M4ZsaERCICLdxLU0EsO7GkKQTNoxm9uRumuXYtWqTJA/Xco/f05la4udNT2g70s0Z/VqdiOtgL0+lp5C/xadrlIkXp+ukZfkziQdYCMpEtNsOUgwdv/Q7Sy9eWHIXXBtju7fMrqH3WRPCkAfsb0B5P1SkJTIWYVYhWQGKta1mWydWsFqnI1HdDmla+rNMEinIcF8e+jHH9XzMzlpgSvt+J07MjLj1z7UsI0xx8m3U9mtepxXIBcWZ5TqdZlu/rNMfyA53mWZ7X6QhLW6ejLD/UaYHlRzodY3lBC5p038GQizDkAg6QMISlA0NYXoIhLBUMYbkIQ1gWYQjLJRjC8mMYwnIZhrC8rGXV1FNJ49qZWAZsQmBijh65zEXlaiq5VEK7aFRqQ54SbpVUFM+qf2WgXjzyhjmwFkiXyJpfMc6Vj0bl+NYVLW8aO1fAsepvH472OfFS1ouFPwX/1dZUJb1izcOTq/Abhp5sJ6o2qXh0TZfPVT26/l9UVFgL9BtIhVgoyrJscGcihI86nYZqoJVDzGzMPLTrdcuan8P9NzFCFlD9+DcUGgvcg05ZSVnt4KzV19uy3DuDcjgTLEkxN/P6VvgiI7PSfpFZyp6PfB5wBYxKZdhqA60VvNknMQ+Z3iTPBHFbUTZI2tjOBIkNHPOAefOdBCZh6qoN5E7hhg34BWFuwXknXKJ6oyyH7kXs8yik/Fun4kT2qGiMwLPZG2Gv70LKb3EMJDT5pX4MVBWhqRg1FdA0Um6oBl/G2bptQsYO9CMqdsOyrOLDxxb3lZJtGYR8pIjVo6Of1l6iTqrcfmYUl++dvgXBIDUxf3vfdHGQyrtayTJHbQNTtxqVU9eaQ+NVh+rmUfW94+wTOWuabronHnpf06rbwcVcLLD2bQ7SUiYX1PVhhQ2iy8WlUOplNEnvuAcYFhjQ71CKjf+r+th8nitVhdFxJN9O1LfR52AM/A/Yf0f1A9D3Y+hyDS7P95oTn2704WyZrqIX66foNzBrrblZugbc0HQD4iFHrY64yg18pwZxeqS5HOkh4GPdFeIBwCaAxeAT3bWM5lMAo/mMOT7A58xh0GQOgy3mMNhmzhrADnMY7DKHwR5zGHzBnHWAL5nDIGQOg4g5DJ4wJwB4yhwGXzGHwdfMYfANc+4DfMscBjFzGCTMYbCv6dYwzC1e0F2gtkFVoANTT1jcw+JQU2XI/o4Xhv29Qcz+wSCm/qjp9pD6Ey8M9WeDmPqLQUz9VdOdIfU3Xhjq7wYx9Q+DmPpMvxjLZQa/jHyXCgeUXWw+5++J9w/bxUC5AAEAAf//AA94nIVVX2hbZRQ/5/t7893s5ja9f7ouzdZ0TTqz3bRJmogbWya6bG6Cq0VbSV2ddIJjFtfIQHEig80Hda8yUN/0YQz8AyriiyD+xQd92R4HCnaCb3samnpumrpsCsLlfPf7zvedc37nL3CAtc/5W/wQZGA3tOBSY/g+TMjHmwzEoM1Q8+ZjRZY4oJhmBw5/YB6Za0yC5AkhlwA1A1yCBIBOwCII0Cj0U8BAMdUCzq05sKwkP7SlUY6fcJk4Fb/RyE79/6P5hjM/F4aZiXBoeMgzcqQ4Xi1hPqfDLG5FT+lchCVU3lYMyvuwhl1mqndQL0RsuloLywHtthLXI06OblTrhfWVnpSJ5+mwu/JdbtuN3IAnkW0LLMcRwaC7ktrlzridM6kVdyf9uO1UNBByI7JhwtG2sEwab07ORBeilWhqavJCqV0qzZTOl/7ZXQ5TbTcdcFelyGhhRDAQpdqp1FEX3w3cFTc1k9pJQkmm4ySCbSikxRP2QOfN+0tHS5MrpQuTU1Mk5nw0E5Xa0WvrOwDyGax9yB9ma6DAg82wHc43SAGTI4GjBWebOePAERFE8/AHaQpZASSTy8A4WwZiLQMQ82mFKATO0ILicRAoDm9p5P99E5b/fXG+kQYY3TYUuqmERWYoT0u/GNYL2q/4WB3LaVS+VynXsVYIcWw6DkCh3nX1D+VzlYN4LClF5yexSQos8exqZ3KVP+wtrC54u4Nznq6cq+xpMpUUnZ8FUYzE86ud0g28NOIv3Gj5/rmA3ABs7S/ywzFuQ4qyd6QxfNtiQIaEgp3w/entQg4Vcbqa16M5FfpeUB8t1+qeg7mI7cUyOe79wOk86gSxkVec4KPTX69++5x68Yubn5/F+w52z7u08sJX7fZXv8ekT/d2mILJxq6sn+SC6qEJknzLJCxyZEKwWVqYmAPBxBE/9DLeZiWHu7lcr/VytrCRuHojncNuTt9h46tmacmYisnSamdN2bZptcsmSysdVsy1PrOvOzF3xN64Rb937t/og9KHxYdcjIUqFAmIAHGHNzlns+RTPgeUYAQm9DwpNxfxbhhBHPaw3/gfTcXO2L+eJVIx5nsyGkvm9X4/f+bGkH45G0PaSjcMXTjcZyTvi3UdHoCDjQd3IDUVsgwYmUoJK/gp4JJxeRI0MKHZIkgynyIBqBTOUs6rOVCojvjZ4mCQz49ZMlMcp8QoYk6NoBfsxnJtsBohpa8iGJS+ZH7gU7NxME6cmF+t7cO9vB8d3jTWSct0ycW9ranXmolNDwmVkNnxe+8JtoztwS5rKJ0xWS95tQ/1zMYzg69MzUZnNtl1ofNbsml/OJm6f9wjRjpnu2o4MzHzn77IQkRd+1DjwMQ2pqSjGMMhyjrgTbBAKksuUm0iU7hI0aN2wOKOq7WYBSH0HGihj/jkiPxAfmwsEbfYrjMG+j3ij932Db/LV7I/xruNrhnroxjR9HRMb2nTvO0ZXOoHPk8H2ZhDPx93qcE/53sH5np/dkIP7zzhTVKdR/BAY/9ElkkR+A6lJGsqpJ4oQcTxpvBT3Kn58VkaJjgHyPEIws57xkaHh9KuVpDEpJZeMbZ5w/zBHi5NMQ4r5VphsFqID7TyB9eR4pX216c3AHxpdAwoqU9qg0ZJ6yVLKmMSz1iG2z27ifx18NkY0LPx1W/wCc2l5LrznrIsiKsqbmB78A9wIGx4tI8rjihVHJyY9pgMirenVq0yWg7Iw7eogG7ZgYM3qR9959A/fZkg6MnD/exlkmc+jWV4SB15XUR+eqC6l6ZmgPtN9z5JMfik05OV8ljylunJ4J+wA/FUaQSSKotsYsCWqaPBidBLcxkWx7XKFRIb45TGaEhjlF9uUVPqXOtcIwsXbBvfoZXIyRYFdkfnqjExH98xpnPczqzjX/uNdO1Y17Wpi5+6Ts8BXtjVFasp9KZ1mOiNbH65c5w6HgmyF2jFCZywM8mWjRc7T5Pmt0lRy7Y71+jYbpGyvwG4sH0XeJxjYGRgYADiwBB/53h+m68M3MwvgCIM1z5N/g6j///9v5H5BbMnkMvBwAQSBQCIcA9gAHicY2BkYGAO+p8FJF/8//v/F/MLBqAICuAFALYQB5kAeJxjfsHAwLwAiCNB+P9fbJjJmoGBMRUo/wKCAfO2EnQAAAAAANoBXgGcAgICVALaA1IDvAPkBAYEPARyAAEAAAANAF0ABAAAAAAAAgAUACQAcwAAAG4LcAAAAAB4nHWRzWrCQBSFT+pPqUIXLXTTzayKUohGKIibCoLuhbrrYtTRxCYZmYyKyz5Fd32HvlDfoO/QkziIFJtw9bvnnpl7ZwLgBt/wcHieGAf2UGd24Atcou+4RH3kuEweO66QXx1XyaHjGh6ROa7jFp/cwStfMVvhy7GHO+/e8QWuvcBxifqz4zL5xXGF/Oa4Sn53XMPE+3Bcx4P3M9DrvYmWoRWNQVN02kFXTPdCU4pSGQu5saE2meiLhU6timPtz3SSs9ypTCdqrJabWJoT5QQnymSRTkXgt0/UkUqVkVbN807ZdtmxdiEWRidi6HqItdErNbN+aO2612qd9sYAGmvsYRBhyUu0EGhQbfK/gzYCdElTOgSdB1eEFBIxFYkNV4RFJWPeZyyYpVQVHTHZx4y/yVGX2LGWFZri51TccUOn5B7nPefVCSPvGhVVwUl9znveO2KkhV8Wk82PZ8qwZf8OVcu1+fSmWCMw/HMOwXvKaysqM+p+cVuWag8tvv+c+xdd+4+teJxtjUEOwiAURJla24KliQfhUA2g/Sl+CKXx+loNrpzVezOLEY34Ron/0WhwQoszOvQYIKFwwQiNSbSBeO2SZ0tBP4j3zVjKNng32ZmtD1VVXCuOiw/pJ8S3WOU6l+K5UOTaDC4+2TjKMtN9KQf1ezLx/Sg/00FCvABHhjDjAAB4nGPw3sFwIihiIyNjX+QGxp0cDBwMyQUbGVidNjEwMmiBGJu5mBg5ICw+BjCLzWkX0wGgNCeQze60i8EBwmZmcNmowtgRGLHBoSNiI3OKy0Y1EG8XRwMDI4tDR3JIBEhJJBBs5mFi5NHawfi/dQNL70YmBhcADHYj9AAA) format('woff');
}

.markdown-body {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #333333;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body b,
.markdown-body strong {
  font-weight: bold;
}

.markdown-body mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

.markdown-body sub,
.markdown-body sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
.markdown-body sup {
  top: -0.5em;
}
.markdown-body sub {
  bottom: -0.25em;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre,
.markdown-body samp {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body .codehilitetable,
.markdown-body .highlighttable {
  border: 0;
  border-spacing: 0;
}

.markdown-body .codehilitetable tr,
.markdown-body .highlighttable {
  border: 0;
}

.markdown-body .codehilitetable pre,
.markdown-body .codehilitetable div.codehilite,
.markdown-body .highlighttable pre,
.markdown-body .highlighttable div.highlight {
  margin: 0;
}

.markdown-body .linenos,
.markdown-body .code,
.markdown-body .codehilitetable td,
.markdown-body .highlighttable td {
  border: 0;
  padding: 0;
}

.markdown-body td:not(.linenos) .linenodiv {
  padding: 0 !important;
}

.markdown-body .code {
  width: 100%;
}

.markdown-body .linenos div pre,
.markdown-body .linenodiv pre,
.markdown-body .linenodiv {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-left-radius: 3px;
  -webkit-border-bottom-left-radius: 3px;
  -moz-border-radius-topleft: 3px;
  -moz-border-radius-bottomleft: 3px;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.markdown-body .code div pre,
.markdown-body .code div {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-right-radius: 3px;
  -webkit-border-bottom-right-radius: 3px;
  -moz-border-radius-topright: 3px;
  -moz-border-radius-bottomright: 3px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.markdown-body * {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body input {
  font: 13px Helvetica, arial, freesans, clean, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol";
  line-height: 1.4;
}

.markdown-body a {
  color: #4183c4;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:focus,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before,
.markdown-body hr:after {
  display: table;
  content: " ";
}

.markdown-body hr:after {
  clear: both;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre,
.markdown-body samp {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #e7e7e7;
  background-image: -moz-linear-gradient(#fefefe, #e7e7e7);
  background-image: -webkit-linear-gradient(#fefefe, #e7e7e7);
  background-image: linear-gradient(#fefefe, #e7e7e7);
  background-repeat: repeat-x;
  border-radius: 2px;
  border: 1px solid #cfcfcf;
  color: #000;
  padding: 3px 5px;
  line-height: 10px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  display: inline-block;
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body .headerlink {
  font: normal 400 16px fontawesome-mini;
  vertical-align: middle;
  margin-left: -16px;
  float: left;
  display: inline-block;
  text-decoration: none;
  opacity: 0;
  color: #333;
}

.markdown-body .headerlink:focus {
  outline: none;
}

.markdown-body h1 .headerlink {
  margin-top: 0.8rem;
}

.markdown-body h2 .headerlink,
.markdown-body h3 .headerlink {
  margin-top: 0.6rem;
}

.markdown-body h4 .headerlink {
  margin-top: 0.2rem;
}

.markdown-body h5 .headerlink,
.markdown-body h6 .headerlink {
  margin-top: 0;
}

.markdown-body .headerlink:hover,
.markdown-body h1:hover .headerlink,
.markdown-body h2:hover .headerlink,
.markdown-body h3:hover .headerlink,
.markdown-body h4:hover .headerlink,
.markdown-body h5:hover .headerlink,
.markdown-body h6:hover .headerlink {
  opacity: 1;
  text-decoration: none;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre,
.markdown-body .admonition {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body code,
.markdown-body samp {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  border-radius: 3px;
}

.markdown-body code:not(.highlight):not(.codehilite), .markdown-body samp {
  background-color: rgba(0,0,0,0.04);
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .codehilite,
.markdown-body .highlight {
  margin-bottom: 16px;
}

.markdown-body .codehilite pre,
.markdown-body .highlight pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
}

.markdown-body .codehilite,
.markdown-body .highlight,
.markdown-body pre {
  border-radius: 3px;
}

.markdown-body :not(.highlight) > pre {
  background-color: #f7f7f7;
}

.markdown-body .codehilite pre,
.markdown-body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

/* Admonition */
.markdown-body .admonition {
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  position: relative;
  border-radius: 3px;
  border: 1px solid #e0e0e0;
  border-left: 6px solid #333;
  padding: 10px 10px 10px 30px;
}

.markdown-body .admonition table {
  color: #333;
}

.markdown-body .admonition p {
  padding: 0;
}

.markdown-body .admonition-title {
  font-weight: bold;
  margin: 0;
}

.markdown-body .admonition>.admonition-title {
  color: #333;
}

.markdown-body .attention>.admonition-title {
  color: #a6d796;
}

.markdown-body .caution>.admonition-title {
  color: #d7a796;
}

.markdown-body .hint>.admonition-title {
  color: #96c6d7;
}

.markdown-body .danger>.admonition-title {
  color: #c25f77;
}

.markdown-body .question>.admonition-title {
  color: #96a6d7;
}

.markdown-body .note>.admonition-title {
  color: #d7c896;
}

.markdown-body .admonition:before,
.markdown-body .attention:before,
.markdown-body .caution:before,
.markdown-body .hint:before,
.markdown-body .danger:before,
.markdown-body .question:before,
.markdown-body .note:before {
  font: normal normal 16px fontawesome-mini;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 1.5;
  color: #333;
  position: absolute;
  left: 0;
  top: 0;
  padding-top: 10px;
  padding-left: 10px;
}

.markdown-body .admonition:before {
  content: "\f056\00a0";
  color: 333;
}

.markdown-body .attention:before {
  content: "\f058\00a0";
  color: #a6d796;
}

.markdown-body .caution:before {
  content: "\f06a\00a0";
  color: #d7a796;
}

.markdown-body .hint:before {
  content: "\f05a\00a0";
  color: #96c6d7;
}

.markdown-body .danger:before {
  content: "\f057\00a0";
  color: #c25f77;
}

.markdown-body .question:before {
  content: "\f059\00a0";
  color: #96a6d7;
}

.markdown-body .note:before {
  content: "\f040\00a0";
  color: #d7c896;
}

.markdown-body .admonition::after {
  content: normal;
}

.markdown-body .attention {
  border-left: 6px solid #a6d796;
}

.markdown-body .caution {
  border-left: 6px solid #d7a796;
}

.markdown-body .hint {
  border-left: 6px solid #96c6d7;
}

.markdown-body .danger {
  border-left: 6px solid #c25f77;
}

.markdown-body .question {
  border-left: 6px solid #96a6d7;
}

.markdown-body .note {
  border-left: 6px solid #d7c896;
}

.markdown-body .admonition>*:first-child {
  margin-top: 0 !important;
}

.markdown-body .admonition>*:last-child {
  margin-bottom: 0 !important;
}

/* progress bar*/
.markdown-body .progress {
  display: block;
  width: 300px;
  margin: 10px 0;
  height: 24px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #ededed;
  position: relative;
  box-shadow: inset -1px 1px 3px rgba(0, 0, 0, .1);
}

.markdown-body .progress-label {
  position: absolute;
  text-align: center;
  font-weight: bold;
  width: 100%; margin: 0;
  line-height: 24px;
  color: #333;
  text-shadow: 1px 1px 0 #fefefe, -1px -1px 0 #fefefe, -1px 1px 0 #fefefe, 1px -1px 0 #fefefe, 0 1px 0 #fefefe, 0 -1px 0 #fefefe, 1px 0 0 #fefefe, -1px 0 0 #fefefe, 1px 1px 2px #000;
  -webkit-font-smoothing: antialiased !important;
  white-space: nowrap;
  overflow: hidden;
}

.markdown-body .progress-bar {
  height: 24px;
  float: left;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #96c6d7;
  box-shadow: inset 0 1px 0 rgba(255, 255, 255, .5), inset 0 -1px 0 rgba(0, 0, 0, .1);
  background-size: 30px 30px;
  background-image: -webkit-linear-gradient(
    135deg, rgba(255, 255, 255, .4) 27%,
    transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%,
    transparent 77%, transparent
  );
  background-image: -moz-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -ms-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -o-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
}

.markdown-body .progress-100plus .progress-bar {
  background-color: #a6d796;
}

.markdown-body .progress-80plus .progress-bar {
  background-color: #c6d796;
}

.markdown-body .progress-60plus .progress-bar {
  background-color: #d7c896;
}

.markdown-body .progress-40plus .progress-bar {
  background-color: #d7a796;
}

.markdown-body .progress-20plus .progress-bar {
  background-color: #d796a6;
}

.markdown-body .progress-0plus .progress-bar {
  background-color: #c25f77;
}

.markdown-body .candystripe-animate .progress-bar{
  -webkit-animation: animate-stripes 3s linear infinite;
  -moz-animation: animate-stripes 3s linear infinite;
  animation: animate-stripes 3s linear infinite;
}

@-webkit-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@-moz-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

.markdown-body .gloss .progress-bar {
  box-shadow:
    inset 0 4px 12px rgba(255, 255, 255, .7),
    inset 0 -12px 0 rgba(0, 0, 0, .05);
}

/* MultiMarkdown Critic Blocks */
.markdown-body .critic_mark {
  background: #ff0;
}

.markdown-body .critic_delete {
  color: #c82829;
  text-decoration: line-through;
}

.markdown-body .critic_insert {
  color: #718c00 ;
  text-decoration: underline;
}

.markdown-body .critic_comment {
  color: #8e908c;
  font-style: italic;
}

.markdown-body .headeranchor {
  font: normal normal 16px fontawesome-mini;
  line-height: 1;
  display: inline-block;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.headeranchor:before {
  content: '\e157';
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 4px 0.25em -20px;
  vertical-align: middle;
}

.markdown-body diagram-div, .markdown-body div.uml-sequence-diagram, .markdown-body, div.uml-flowchart {
  overflow: auto;
}

/* Media */
@media only screen and (min-width: 480px) {
  .markdown-body {
    font-size:14px;
  }
}

@media only screen and (min-width: 768px) {
  .markdown-body {
    font-size:16px;
  }
}

@media print {
  .markdown-body * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
    -ms-filter: none !important;
  }

  .markdown-body {
    font-size:12pt;
    max-width:100%;
    outline:none;
    border: 0;
  }

  .markdown-body a,
  .markdown-body a:visited {
    text-decoration: underline;
  }

  .markdown-body .headeranchor-link {
    display: none;
  }

  .markdown-body a[href]:after {
    content: " (" attr(href) ")";
  }

  .markdown-body abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .markdown-body .ir a:after,
  .markdown-body a[href^="javascript:"]:after,
  .markdown-body a[href^="#"]:after {
    content: "";
  }

  .markdown-body pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  .markdown-body pre,
  .markdown-body blockquote {
    border: 1px solid #999;
    padding-right: 1em;
    page-break-inside: avoid;
  }

  .markdown-body .progress,
  .markdown-body .progress-bar {
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    box-shadow: none;
  }

  .markdown-body .progress {
    border: 1px solid #ddd;
  }

  .markdown-body .progress-bar {
    height: 22px;
    border-right: 1px solid #ddd;
  }

  .markdown-body tr,
  .markdown-body img {
    page-break-inside: avoid;
  }

  .markdown-body img {
    max-width: 100% !important;
  }

  .markdown-body p,
  .markdown-body h2,
  .markdown-body h3 {
    orphans: 3;
    widows: 3;
  }

  .markdown-body h2,
  .markdown-body h3 {
    page-break-after: avoid;
  }
}
</style><style>/*GitHub*/
.highlight {background-color:#f7f7f7;color:#333333;}
.highlight .hll {background-color:#ffffcc;}
.highlight .c{color:#999988;font-style:italic}
.highlight .err{color:#a61717;background-color:#e3d2d2}
.highlight .k{font-weight:bold}
.highlight .o{font-weight:bold}
.highlight .cm{color:#999988;font-style:italic}
.highlight .cp{color:#999999;font-weight:bold}
.highlight .c1{color:#999988;font-style:italic}
.highlight .cs{color:#999999;font-weight:bold;font-style:italic}
.highlight .gd{color:#000000;background-color:#ffdddd}
.highlight .ge{font-style:italic}
.highlight .gr{color:#aa0000}
.highlight .gh{color:#999999}
.highlight .gi{color:#000000;background-color:#ddffdd}
.highlight .go{color:#888888}
.highlight .gp{color:#555555}
.highlight .gs{font-weight:bold}
.highlight .gu{color:#800080;font-weight:bold}
.highlight .gt{color:#aa0000}
.highlight .kc{font-weight:bold}
.highlight .kd{font-weight:bold}
.highlight .kn{font-weight:bold}
.highlight .kp{font-weight:bold}
.highlight .kr{font-weight:bold}
.highlight .kt{color:#445588;font-weight:bold}
.highlight .m{color:#009999}
.highlight .s{color:#dd1144}
.highlight .n{color:#333333}
.highlight .na{color:teal}
.highlight .nb{color:#0086b3}
.highlight .nc{color:#445588;font-weight:bold}
.highlight .no{color:teal}
.highlight .ni{color:purple}
.highlight .ne{color:#990000;font-weight:bold}
.highlight .nf{color:#990000;font-weight:bold}
.highlight .nn{color:#555555}
.highlight .nt{color:navy}
.highlight .nv{color:teal}
.highlight .ow{font-weight:bold}
.highlight .w{color:#bbbbbb}
.highlight .mf{color:#009999}
.highlight .mh{color:#009999}
.highlight .mi{color:#009999}
.highlight .mo{color:#009999}
.highlight .sb{color:#dd1144}
.highlight .sc{color:#dd1144}
.highlight .sd{color:#dd1144}
.highlight .s2{color:#dd1144}
.highlight .se{color:#dd1144}
.highlight .sh{color:#dd1144}
.highlight .si{color:#dd1144}
.highlight .sx{color:#dd1144}
.highlight .sr{color:#009926}
.highlight .s1{color:#dd1144}
.highlight .ss{color:#990073}
.highlight .bp{color:#999999}
.highlight .vc{color:teal}
.highlight .vg{color:teal}
.highlight .vi{color:teal}
.highlight .il{color:#009999}
.highlight .gc{color:#999;background-color:#EAF2F5}
</style><title>TE_tutorial_1.1</title></head><body><article class="markdown-body"><h1 id="introduction-to-transposable-element-annotation">Introduction to transposable element annotation<a class="headerlink" href="#introduction-to-transposable-element-annotation" title="Permanent link"></a></h1>
<blockquote>
<p>2025 Workshop on Genomics, Cesky Krumlov - 16th January 2025</p>
</blockquote>
<p><strong>Valentina Peona</strong></p>
<p><em><a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#118;&#97;&#108;&#101;&#110;&#116;&#105;&#110;&#97;&#46;&#112;&#101;&#111;&#110;&#97;&#64;&#110;&#114;&#109;&#46;&#115;&#101;">&#118;&#97;&#108;&#101;&#110;&#116;&#105;&#110;&#97;&#46;&#112;&#101;&#111;&#110;&#97;&#64;&#110;&#114;&#109;&#46;&#115;&#101;</a></em></p>
<h2 id="table-of-contents">Table of contents<a class="headerlink" href="#table-of-contents" title="Permanent link"></a></h2>
<ul>
<li><a href="#Introduction">Introduction</a></li>
<li><a href="#RepeatModeler2">Run RepeatModeler2</a></li>
<li><a href="#RepeatMasker">Run RepeatMasker</a></li>
<li><a href="#RepeatLandscape">Create a repeat landscape</a></li>
<li><a href="#Databases">TE databases</a></li>
<li><a href="#TEHub">TE Hub</a></li>
<li><a href="#TEdiversity">Exploration of TE diversity</a></li>
</ul>
<h2 id="introduction">Introduction <a name="introduction"></a><a class="headerlink" href="#introduction" title="Permanent link"></a></h2>
<p>The characterisation of the diversity and location of transposable elements in a genome can be key for many analyses. Sometimes it is important to know their sequences and where they are to directly study them, sometimes it is important to know their coordinates to avoid them! For example, it is necessary to get rid of transposable elements (or repetitive elements in general) when annotating host genes or when refining your SNP calls and running popgen-related downstream analyses (e.g., nucleotide diversity, selection scans).</p>
<p>Whatever the reason you look at TEs, everything starts with the:</p>
<p>1) <strong>characterisation of the diversity</strong> of repetitive elements in your genome of interest by creating a de-novo library of consensus sequences or by retrieving the consensus sequences from databases. We are going to create a de-novo library using <a href="https://www.pnas.org/doi/10.1073/pnas.1921046117">RepeatModeler2</a>.</p>
<p>2) <strong>annotation</strong> of the repetitive elements, namely the identification of the coordinates of the repeats in a genome and their sequence nature (also referred to as &ldquo;genome masking&rdquo;) using a library of known TEs. We are going to use <a href="https://www.repeatmasker.org/">RepeatMasker</a> for this task.</p>
<hr />
<p>This tutorial will guide you through the use of these two softwares and at the end of this activity you should know how to:</p>
<p align="center"> 📖 <strong>Learning objectives</strong> 📖 </p>

<ul class="task-list">
<li class="task-list-item"><input type="checkbox" disabled/> create of a de-novo repeat library using RepeatModeler2</li>
<li class="task-list-item"><input type="checkbox" disabled/> describe what a repeat library and consensus sequences are</li>
<li class="task-list-item"><input type="checkbox" disabled/> use a repeat library to annotate repeats in a genome with RepeatMasker</li>
<li class="task-list-item"><input type="checkbox" disabled/> find already characterised consensus sequences in a database for your species of interest</li>
<li class="task-list-item"><input type="checkbox" disabled/> explore some key sequence characteristics of TEs to distinguish the main TE categories</li>
</ul>
<hr />
<h2 id="1-de-novo-creation-of-a-repeat-library-repeatmodeler2">1) De-novo creation of a repeat library - RepeatModeler2 <a name="RepeatModeler2"></a><a class="headerlink" href="#1-de-novo-creation-of-a-repeat-library-repeatmodeler2" title="Permanent link"></a></h2>
<p>Let&rsquo;s start from the beginning and create a library of representative sequences (consensus sequences) for all the types of repetitive elements present in our genome of interest.</p>
<p><img align="absmiddle" alt=":memo:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/1f4dd.png" title=":memo:" width="20px" /> <strong>Libraries and consensus sequences</strong> <img align="absmiddle" alt=":memo:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/1f4dd.png" title=":memo:" width="20px" /></p>
<blockquote>
<p>A TE family (that can be seen as a species) can be represented by a consensus sequence approximating that of the ancestral progenitor. Such consensus sequence can be created on top of a multiple alignment of individual genomic copies (or “seeds”) from which each ancestral nucleotide can be inferred based on a majority rule along the alignment. Similarly, the seed alignment may be used to generate a profile Hidden Markov Model (HMM) for each family. <a href="https://www.pnas.org/doi/10.1073/pnas.1921046117">Flynn et al., 2020</a></p>
</blockquote>
<p>A library of consensus sequences will then be essential to find sequences in the genome of interest that are similar to the consensus sequences themselves, namely the different copies of the TEs.</p>
<p>A good repeat library has these characteristics:</p>
<p>🟢 is complete - the entire diversity of repeats is represented</p>
<p>🟢 contains nonredundant consensus sequences - each element is represented only once</p>
<p>🟢 contains full-length consensus sequences - each elements is not fragmented/truncated</p>
<p>If you work on model organisms, it is very likely that the diversity of transposable elements of your species has already been characterised and that you can find a library of consensus sequences avaialable on databases like <a href="https://www.girinst.org/repbase/">Repbase</a> and <a href="https://dfam.org/">Dfam</a>. In this case, it is not essential to run tools like RepeatModeler2. This can be true also if your species of interest is closely related to a species whose transposable elements have been already characterised. However, we find over and over that masking a genome of a new species using repeat libraries from other species can be insufficient. This happens when the genome of interest contains species-specific repetitive elements never described before and these elements would remain largely unmasked or partially and/or incorrectly annotated. Furthermore, curated repeat libraries from sister species can actually reciprocally help the annotation of the two genomes <a href="https://www.mdpi.com/2073-4425/10/4/301">Boman et al. 2018</a>.</p>
<p>There are a few tools available for the de-novo characterisation of repeats (e.g., CARP, REPET, EarlGrey, RepeatExplorer2 and other all listed in <a href="https://tehub.org">https://tehub.org</a>) but for time reasons we will use only RepeatModeler2. As target genome, we chose to use the assembled chromosomes of the wheatear <em>Oenanthe melanoleuca</em> (GCA_029582105.1) <a href="https://academic.oup.com/g3journal/article/13/6/jkad088/7141353">Peona et al. 2023</a>. </p>
<h4 id="lets-run-repeatmodeler2">Let&rsquo;s run RepeatModeler2!!<a class="headerlink" href="#lets-run-repeatmodeler2" title="Permanent link"></a></h4>
<p><img align="absmiddle" alt=":warning:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/26a0.png" title=":warning:" width="20px" /> <strong>Warning:</strong> Given the short time we have today, you will not be able to fully run the RepeatModeler2 commands, therefore we generated all the output for you that can be found in the folder <code>~/workshop_materials/transposable_elements/Outputs</code>. I suggest you to follow the tutorial and try the commands out but kill the analysis right after. All the output necessary is available to you in any moment!</p>
<p>A run of RepeatModeler2 on a genome consists of two steps:</p>
<ul>
<li>indexing of the genome (somewhat similar to what bwa index does) using the command <code>BuildDatabase</code></li>
<li>clustering of similar sequences to create the de-novo library using the command <code>RepeatModeler</code></li>
</ul>
<p align="center"><figure><img src="./repeatmodeler2.jpg" alt="RepeatModeler2 pipeline" /><figcaption><strong>Figure 1.</strong> RepeatModeler2 pipeline. Image from <a href="https://www.pnas.org/doi/10.1073/pnas.1921046117">Flynn et. al 2020</a>.</figcaption></figure></p>

<p>The core of the RepeatModeler2 pipeline (illustrated above in <strong>Figure 1</strong>) consists of up to 6 rounds of sequence subsampling of your genome. Each round subsamples chunks of your genome of different sizes and on each of these chunks, several tools are run to identify both tandem and interspersed repeats. At the end of each round, multisequence alignments of repetitive elements are created and consensus sequences are produced on top of them and stored in the <code>consensi.fa</code> file. RepeatModeler2 can also run an optional additional pipeline called LTRstruct particularly dedicated to improve the characterisation of LTR retrotransposons.</p>
<p>Before running the first commands, let&rsquo;s set some variables for the most used files and paths</p>
<div class="highlight"><pre><span class="c1"># activate conda environment</span>
conda activate transposable_elements

<span class="nv">WD</span><span class="o">=</span>~/workshop_materials/transposable_elements
<span class="nv">OUTDIR</span><span class="o">=</span><span class="nv">$WD</span>/Results
<span class="nv">REF</span><span class="o">=</span><span class="nv">$WD</span>/Genomes/oenMel.fasta

<span class="c1"># go to working directory</span>
<span class="nb">cd</span> <span class="nv">$WD</span>

<span class="c1"># create output folders</span>
mkdir -p <span class="nv">$OUTDIR</span>/RepeatModeler2 <span class="nv">$OUTDIR</span>/RepeatMasker
</pre></div>

<p>Call the command <code>BuildDatabase</code>, look at the options and run it on <code>oenMel.fasta</code> genome.</p>
<div class="highlight"><pre>BuildDatabase
</pre></div>

<details>
    <summary><strong>BuildDatabase command</strong></summary>

<div class="highlight"><pre><span class="c1"># BuildDatabase takes about 1 minute to run</span>
<span class="nb">cd</span> <span class="nv">$WD</span>

BuildDatabase -name <span class="nv">$REF</span> <span class="nv">$REF</span>
</pre></div>

</details>

<p>Now that your index/database is ready, we can run RepeatModeler2 on it.</p>
<p>As before, call the <code>RepeatModeler</code> command and look at the options. Which options would you use?</p>
<p>Below you can find how I ran RepeatModeler2 on <code>oenMel.fasta</code>.</p>
<details>
    <summary><strong>RepeatModeler command</strong></summary>

<div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$WD</span>

<span class="c1"># Run RepeatModeler2 with the LTR pipeline</span>
RepeatModeler -database <span class="nv">$REF</span> -threads <span class="m">8</span> -LTRStruct
</pre></div>

</details>

<p><br>
<img align="absmiddle" alt=":warning:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/26a0.png" title=":warning:" width="20px" /> <strong>Warning:</strong> Once you managed to start RepeatModeler2, let it run for a minute or two (or until you lose your patience) to see the very first round of repeat prediction start and then please kill the command with <code>Ctrl + C</code> as we do not have the time to let it finish. You will see that a folder with a name similar to this <code>RM_1962636.MonJan131517222025</code> has been created in your working directory that contains the intermediate files of the RepeatModeler2 pipeline. You can give it a look and then you can jump directly to the final output files of the pipeline.</p>
<p><img align="absmiddle" alt=":exclamation:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2757.png" title=":exclamation:" width="20px" /> Copy the output from the <code>Outputs</code> folder to your <code>Results</code> folder:</p>
<div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$WD</span>
cp -r <span class="nv">$WD</span>/Outputs/RepeatModeler2 <span class="nv">$OUTDIR</span>/
</pre></div>

<p>📖 <strong>Let&rsquo;s give a look at the output of RepeatModeler2</strong> 📖</p>
<p>Go to your <code>Results</code> folder and list the files present there.</p>
<p><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /> <strong>Questions</strong> <img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /></p>
<p>🔴 What files do you find? What is the difference between a fasta file and a Stockholm file?</p>
<p>🔴 What is the difference between the classified and non-classified fasta files?</p>
<p>🔴 What is a consensus sequence?</p>
<p>🔴 Do you see shortcomings in the use of consensus sequences to annotate repeats?</p>
<p>🔴 Do you think this library is complete, non-redundant and contains full-length elements? (Just reflect on this, it is difficult to know at this stage.)</p>
<p>🔴 Do you think this library contains only transposable elements?</p>
<details>
    <summary><strong>Answers</strong></summary>

<strong>What files do you find?</strong>
In the <code>Results</code> folder you find the final output of RepeatModeler2. Note that the intermediate files have been removed. RepeatModeler2 creates a huge amount of intermediate files that are useful essentially to resume a run that failed.
<br>
The fasta and Stockholm files contain the consensus sequences generated by the tool. While the fasta file can be directly used by RepeatMasker, the Stockholm file can be used to create HMM profiles and be used by other HMM-based tools.
<br>

<div class="highlight"><pre>consensi.fa  : Consensus sequences
consensi.fa.classified: Consensus sequences classified
families.stk : Seed alignments
families-classified.stk: Seed alignments classified
</pre></div>

<br>

<strong>What is the difference between the classified and non-classified fasta files?</strong>
The difference lies in the headers of the fasta files. In the classified version you find a tag after the name of the repeat that specifies (when possible) the type of repeat. Note that this classification can be partial and sometimes incorrect.
<br><br>

<strong>What is a consensus sequence? Do you see shortcomings in the use of consensus sequences to annotate repeats?</strong>
A consensus sequence is an approximation of the ancestral repeat that gave rise to the insertions found in the genome. Note that a consensus sequence is made on top of a set of similar (and likely homologous) repetitive elements, therefore it represents an approximation of the ancestor of that limited set of aligned sequences therefore could not capture the entire variability of that family of TE.
<br><br>

<strong>Do you think this library is complete, non-redundant and contains full-length elements?</strong>
Most likely, the repeat library just produced here does not meet all these criteria. Likely it is incomplete because low copy number repeats are still difficult to identify and build a consensus on top of a very few sequences. Likely some of the consensus sequences are not full-length because of the nature of the transposable elements (e.g., LINEs and 5' truncation) or because of the fragmentation of the genome used. However, the library should rather be non-redundant because RepeatModeler2 has a step in the pipeline that gets rid of most of the redundancy. It is good practice to check for it! <br> A library can be improved with some manual curation!
<br><br>

<strong>Do you think this library contains only transposable elements?</strong> The library likely contains also tandem repeats but it can also contain multi-copy genes. RepeatModeler2 takes care of removing gene-related sequences but it is still possible that some genes remain. You can check for it using blast or tools like <a href="https://github.com/NBISweden/ProtExcluder">ProtExcluder</a>.


</details>

<p><br>
The library produced here contains quite a lot of consensus sequences but possibly it is not of the best quality.</p>
<p>For the scope of this tutorial, we are going to use the library as it is but for more detailed analyses I warmly recommend to curate the library manually with the help of some tools (eg., TE-Aid) or using longer pipelines that build on top of RepeatModeler2 output and automatically curate your consensus sequences. Automatic curation and classification are great but be aware that they are not perfect and at least a general knowledge of the manual curation process is recommended before trusting the result of such pipeline as it is.</p>
<p>:note: Note that the header names of the consensus sequences in the <code>consensi.fa.classified</code> follows a precise nomenclature and style which is necessary for RepeatMasker to give you correct abundance estimates for each kind of TE.</p>
<hr />
<p align="center"> 🔎 <strong>Further readings and tools</strong> 🔎 </p>

<ul>
<li><a href="https://link.springer.com/article/10.1186/s13100-021-00259-7">A beginner’s guide to manual curation of transposable elements</a>. This guide includes several protocols, videos and code.</li>
<li><a href="https://academic.oup.com/mbe/article/41/4/msae068/7635926">Earl Grey: A Fully Automated User-Friendly Transposable Element Annotation and Analysis Pipeline</a>. Example of a tool to get an automatically curated repeat library.</li>
</ul>
<hr />
<h2 id="2-annotation-of-transposable-elements-repeatmasker">2) Annotation of transposable elements - RepeatMasker <a name="RepeatMasker"></a><a class="headerlink" href="#2-annotation-of-transposable-elements-repeatmasker" title="Permanent link"></a></h2>
<p>Now that we have a new repeat library for our genome of interest, we can finally find where the transposable element insertions are and in what abundance with RepeatMasker!</p>
<p>Let&rsquo;s be sure that all our variables are correctly set</p>
<div class="highlight"><pre><span class="nv">LIB</span><span class="o">=</span><span class="nv">$OUTDIR</span>/RepeatModeler2/consensi.fa.classified
</pre></div>

<p>As first thing, let&rsquo;s look at the manual of the tool. To do that just call the <code>RepeatMasker</code> command and then the <code>RepeatMasker --help</code> command (the second prints a very long version of the manual) and read the options. Which options would you use? Below the command I used.</p>
<details>
    <summary><strong>RepeatMasker command</strong></summary>

<div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$WD</span>

<span class="c1"># for a sensitive search -s</span>
RepeatMasker -pa <span class="m">8</span> -s -gff -gccalc -a -excln -lib <span class="nv">$LIB</span> -dir <span class="nv">$OUTDIR</span>/RepeatMasker <span class="nv">$REF</span>

<span class="c1"># for a faster search -q</span>
RepeatMasker -pa <span class="m">8</span> -q -gff -gccalc -a -excln -lib <span class="nv">$LIB</span> -dir <span class="nv">$OUTDIR</span>/RepeatMasker <span class="nv">$REF</span>
</pre></div>

</details>

<p><br>
<img align="absmiddle" alt=":warning:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/26a0.png" title=":warning:" width="20px" /> <strong>Warning:</strong> Once you managed to start RepeatMasker, wait a again a couple of minutes for you to see the analysis start but then please kill the command with <code>Ctrl + C</code> as we do not have the time to let it finish.</p>
<p><img align="absmiddle" alt=":exclamation:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2757.png" title=":exclamation:" width="20px" /> Copy the entire output from the <code>Outputs</code> folder to your <code>Results</code> folder:</p>
<div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$WD</span>
cp -r <span class="nv">$WD</span>/Outputs/RepeatMasker <span class="nv">$OUTDIR</span>/
</pre></div>

<p>Look at the output files of RepeatMasker (especially the <code>.tbl</code> and <code>.out</code> files) and answer the questions.</p>
<p><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /> <strong>Questions</strong> <img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /></p>
<p>🔴 What&rsquo;s the difference between <code>-species</code> and <code>-lib</code> options?</p>
<p>🔴 What do all the columns in the <code>.out</code> file mean? What does divergence mean?</p>
<p>🔴 Can you hypothesise what it means when two or more rows share the same ID number (last column) in the <code>.out</code> file?</p>
<h2 id="create-a-repeat-landscape">Create a repeat landscape<a class="headerlink" href="#create-a-repeat-landscape" title="Permanent link"></a></h2>
<p>Now we can run one little donwstream analysis to visualise the abundance of repeats in your genome of interest using the scripts available from the RepeatMasker tool suite.</p>
<p>We use two Perl scripts. The first <code>calcDivergenceFromAlign.pl</code> uses the <code>.align</code> file to recalculate the divergence of each hit from its consensus sequence and creates a <code>.divsum</code> file which summarises the abundances of each repeat at different percentages of divergence. The <code>.divsum</code> file is then used by the script <code>createRepeatLandscape.pl</code> to create a barplot where you can explore the abundance of each type of repeats found in the genome at different divergence percentages.</p>
<div class="highlight"><pre><span class="nb">cd</span> <span class="nv">$WD</span>

<span class="c1"># it takes about 2 minutes</span>
calcDivergenceFromAlign.pl -s <span class="nv">$OUTDIR</span>/RepeatMasker/oenMel.fasta.divsum <span class="nv">$OUTDIR</span>/RepeatMasker/oenMel.fasta.align

<span class="c1"># it takes a couple of seconds</span>
createRepeatLandscape.pl -div <span class="nv">$OUTDIR</span>/RepeatMasker/oenMel.fasta.divsum -g <span class="m">1200000000</span> &gt; <span class="nv">$OUTDIR</span>/RepeatMasker/oenMel.fasta.html

<span class="c1"># download the html file to your local computer and open it on your browser and admire the TEs &lt;3</span>
</pre></div>

<p><code>-g</code> option requires the size of the genome used, the genome size of our bird is around 1.1 and 1.2 Gb.</p>
<p>If for any reason, the above two Perl scripts do not work for you (they really should!), you can find the output files in <code>Outputs/Landscapes/</code>. Copy the files to your working directory and/or to your local computer.</p>
<p><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /> <strong>Questions</strong> <img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /><img align="absmiddle" alt=":question:" class="gemoji" height="20px" src="https://github.githubassets.com/images/icons/emoji/unicode/2753.png" title=":question:" width="20px" /></p>
<p>🔴 Print the help page for <code>calcDivergenceFromAlign.pl</code>. Can you see what the script takes into consideration to calculate the divergence that RepeatMasker did not before?</p>
<p>🔴 Copy the <code>.html</code> file into your local. Which repeats have been/are likely active in recent time?</p>
<p>🔴 Would you trust the annotation of insertions at very high percentages of divergence? Do you think there is a better way to find them instead of using a homology-based approach like RepeatMasker?</p>
<h2 id="databases-of-known-transposable-element-sequences">Databases of known transposable element sequences <a name="Databases"></a><a class="headerlink" href="#databases-of-known-transposable-element-sequences" title="Permanent link"></a></h2>
<p>At present, a lot of TE libraries are being produced for many different species and uploaded to dedicated databases like <a href="https://www.girinst.org/repbase/update/browse.php">Repbase</a> and Dfam.</p>
<p>I suggest you to go to Repbase or Dfam and look for TE consensus sequences of your species.
Note that unfortunately Repbase requires a subscription to download the sequences but you can still explore them by going to the website and click on the &ldquo;Browse&rdquo; option on the top left of the <a href="https://www.girinst.org/repbase/update/browse.php">page</a>.</p>
<p>Also go to <a href="https://www.dfam.org/home">Dfam</a> and browse for your species! Try to download a consensus sequence from the database, especially try to download the fasta file for a TE of your choice.</p>
<p>🔴 Is the header format suitable for RepeatMasker?</p>
<details>
    <summary><strong>Answer</strong></summary>

Unfortunately, the header given by directly downloading the fasta file from Dfam is not in a RepeatMasker-friendly format. To download a library in the RepeatMasker format, you can use this tool https://github.com/Dfam-consortium/FamDB

</details>

<p><br></p>
<h2 id="te-hub">TE Hub <a name="TEHub"></a><a class="headerlink" href="#te-hub" title="Permanent link"></a></h2>
<p>TE Hub is a community effort to collect all the TE-related resources and tools and make them easily accessible to everyone. I warmly suggest you to look at the <a href="https://www.tehub.org/">website</a> if interested in deepening your knowledge of TEs.</p>
<p>For more info you can also read the associated paper published on <em>Mobile DNA</em>:</p>
<p><a href="https://mobilednajournal.biomedcentral.com/articles/10.1186/s13100-021-00244-0"><strong>TE Hub: A community-oriented space for sharing and connecting tools, data, resources, and methods for transposable element annotation</strong></a></p>
<blockquote>
<p>The diversity of available TE-related methods and resources raises compatibility concerns and can be overwhelming to researchers and communicators seeking straightforward guidance or materials. To address these challenges, we have initiated a new resource, TE Hub, that provides a space where members of the TE community can collaborate to document and create resources and methods. The space consists of (1) a website organized with an open wiki framework, <a href="https://tehub.org">https://tehub.org</a>, (2) a conversation framework via a Twitter account and a Slack channel, and (3) bi-monthly Hub Update video chats on the platform’s development. In addition to serving as a centralized repository and communication platform, TE Hub lays the foundation for improved integration, standardization, and effectiveness of diverse tools and protocols.</p>
</blockquote>
<hr />
<p align="center"><figure><img src="./mobiledna.jpg" alt="Mobile DNA logo" /></figure></p>

<p align="center"> <em>Mobile DNA</em> is the main journal of the TE community. Note that its logo is a DNA transposon ❤️ </p>

<hr />
<h2 id="exploration-of-te-diversity">Exploration of TE diversity <a name="TEdiversity"></a><a class="headerlink" href="#exploration-of-te-diversity" title="Permanent link"></a></h2>
<p align="center"><figure><img src="./teaid.png" alt="TE Aid logo" /></figure></p>

<p>As last part of this tutorial I propose you to give a look at the features of some TE consensus sequences from a couple of organisms (wheatears and tardigrades!) so to get an idea of how to distinguish the different types of TEs and try to classify them. Use the last slides of the lecture where I added some very useful classification tables.</p>
<p>I used <a href="https://github.com/clemgoub/TE-Aid">TE-Aid</a> (which takes our TE consensus sequences as input (fasta file), aligns them to a reference genome) to produce 4 different plots reporting (<strong>Figure 2</strong>):</p>
<ul>
<li>(top left) the genomic hits with divergence to consensus</li>
<li>(top right) the genomic coverage of the consensus</li>
<li>(bottom left) a self dot-plot</li>
<li>(bottom right) a structure analysis including: TIR and LTR suggestions, open reading frames (ORFs) and TE protein hit annotation.</li>
</ul>
<p align="center"><figure><img src="./teaid_plot.jpeg" alt="TE Aid output example"/><figcaption><strong>Figure 2.</strong> TE-Aid example output. Image from <a href="https://mobilednajournal.biomedcentral.com/articles/10.1186/s13100-021-00259-7">Goubert et. al 2022</a>.</figcaption></figure></p>

<p>The plots I produced for you and the corresponding sequences can be downloaded from the Workshop page.</p>
<p>Given today&rsquo;s quick introduction to TEs, the aim of this section is for you to get an idea of which are the important sequence characteristics to distinguish TEs from the 4 broad categories we mentioned during the lecture.</p></article></body></html>